akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java

```diff
@@ -0,0 +1,347 @@
+    this.confidence = 1 - 1 / Math.pow(2, depth);
+    initTablesWith(depth, width, seed);
+  }
+
+  public CountMinSketch(double epsOfTotalCount, double confidence, int seed) {
+    // 2/w = eps ; w = 2/eps
+    // 1/2^depth <= 1-confidence ; depth >= -log2 (1-confidence)
+    this.eps = epsOfTotalCount;
+    this.confidence = confidence;
+    this.width = (int) Math.ceil(2 / epsOfTotalCount);
+    this.depth = (int) Math.ceil(-Math.log(1 - confidence) / Math.log(2));
+    initTablesWith(depth, width, seed);
+  }
+
+  CountMinSketch(int depth, int width, int size, long[] hashA, long[][] table) {
+    this.depth = depth;
+    this.width = width;
+    this.eps = 2.0 / width;
+    this.confidence = 1 - 1 / Math.pow(2, depth);
+    this.hashA = hashA;
+    this.table = table;
+    this.size = size;
+  }
+
+  private void initTablesWith(int depth, int width, int seed) {
+    this.table = new long[depth][width];
+    this.hashA = new long[depth];
+    Random r = new Random(seed);
+    // We're using a linear hash functions
+    // of the form (a*x+b) mod p.
+    // a,b are chosen independently for each hash function.
+    // However we can set b = 0 as all it does is shift the results
+    // without compromising their uniformity or independence with
+    // the other hashes.
+    for (int i = 0; i < depth; ++i) {
+      hashA[i] = r.nextInt(Integer.MAX_VALUE);
+    }
+  }
+
+  /** Referred to as {@code epsilon} in the whitepaper */
+  public double getRelativeError() {
+    return eps;
+  }
+
+  public double getConfidence() {
+    return confidence;
+  }
+
+  private int hash(long item, int i) {
+    long hash = hashA[i] * item;
+    // A super fast way of computing x mod 2^p-1
+    // See http://www.cs.princeton.edu/courses/archive/fall09/cos521/Handouts/universalclasses.pdf
+    // page 149, right after Proposition 7.
+    hash += hash >> 32;
+    hash &= PRIME_MODULUS;
+    // Doing "%" after (int) conversion is ~2x faster than %'ing longs.
+    return ((int) hash) % width;
+  }
+
+  public void add(long item, long count) {
+    if (count < 0) {
+      // Actually for negative increments we'll need to use the median
+      // instead of minimum, and accuracy will suffer somewhat.
+      // Probably makes sense to add an "allow negative increments"
+      // parameter to constructor.
+      throw new IllegalArgumentException("Negative increments not implemented");
+    }
+    for (int i = 0; i < depth; ++i) {
+      table[i][hash(item, i)] += count;
+    }
+    size += count;
+  }
+
+  public void add(String item, long count) {
+    if (count < 0) {
+      // Actually for negative increments we'll need to use the median
+      // instead of minimum, and accuracy will suffer somewhat.
+      // Probably makes sense to add an "allow negative increments"
+      // parameter to constructor.
+      throw new IllegalArgumentException("Negative increments not implemented");
+    }
+    final int[] buckets = MurmurHash.hashBuckets(item, depth, width);
```

Comment:
good idea, also I think we'd be able to recycle this one as we only ever add one item at a time